sstables/checksummed_data_source.cc

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <cstdint>
+
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/temporary_buffer.hh>
+
+#include "types.hh"
+#include "exceptions.hh"
+#include "checksum_utils.hh"
+
+namespace sstables {
+
+extern logging::logger sstlog;
+
+// File data source implementation for SSTables with attached checksum
+// data and no compression
+template <ChecksumUtils ChecksumType>
+class checksummed_file_data_source_impl : public data_source_impl {
+    std::optional<input_stream<char>> _input_stream;
+    const checksum& _checksum;
+    uint64_t _underlying_pos;
+    uint64_t _pos;
+    uint64_t _beg_pos;
+    uint64_t _end_pos;
+public:
+    checksummed_file_data_source_impl(file f, const checksum& checksum,
+            uint64_t pos, size_t len, file_input_stream_options options)
+    : _checksum(checksum)
+    , _pos(pos)
+    , _beg_pos(pos)
+    , _end_pos(pos + len)
+    {
+        // _beg_pos and _end_pos specify positions in the stream.
+        // These are not necessarily aligned on chunk boundaries.
+        // To be able to verify the checksums, we need to translate
+        // them into a range of chunks that contain the given
+        // address range, and open a file input stream to read that
+        // range. The _underlying_pos always points to the current
+        // chunk-aligned position of the file input stream.
+        auto chunk_size = checksum.chunk_size;
+        if (chunk_size == 0 || (chunk_size & (chunk_size - 1)) != 0) {
+            on_internal_error(sstlog, format("Invalid chunk size: {}", chunk_size));
+        }
+        auto start = (_beg_pos / chunk_size) * chunk_size;
+        auto end = (_end_pos / chunk_size + 1) * chunk_size;
+        _input_stream = make_file_input_stream(std::move(f), start, end - start, std::move(options));
```

Comment:
I don't see how _input_stream can ever be disengaged, better to avoid making it optional. You can put the instructions to construct it in a static helper or std::invoke().

Reply:
>I don't see how _input_stream can ever be disengaged, better to avoid making it optional.

Yeah, you're probably right. I borrowed this piece of code from the `compressed_file_data_source_impl` where it was made optional when `close()` was introduced: https://github.com/scylladb/scylladb/commit/4513f8b52c0908e4fb96dd773bd441f146590a76. Not sure why this change was made.

Reply:
How was this resolved?

Reply:
I explained this in the changelog but forgot to mention it here, apologies.

The reason for being an optional in the `compressed_file_data_source_impl` is the early return statement in the constructor: https://github.com/scylladb/scylladb/blob/fc690a60d8b43fe678f3cbb267119c6ea57dd64c/sstables/compress.cc#L363

This is an optimization in case there is nothing to read.

I thought I should port this to the checksummed data source and keep the input stream as an optional.